### PR TITLE
Disable mod "Adaptive Speed" in multiplayer

### LINKED ISSUE #17269: Adaptive speed mod should be disabled in multiplayer
### Discussed in https://github.com/ppy/osu/discussions/17268

<div type='discussions-op-text'>

<sup>Originally posted by **iSimpFor** March 16, 2022</sup>
for regular multiplayer, does the speed of the mod change based on all of the player's mistimings or does it just desync?</div>

### PR DESCRIPTION
**Pull request by frenzibyte**
Closes #17269 

Handles disabling mod client-side. Would probably need to be disabled web-side as well.

Had thought about disabling this in playlists as well, but users will be playing alone in that mode, so I presume it's fine to allow it.

### CODE REVIEW

**Comment by smoogipoo** [comment]
We need to extract this to `ModUtils` for now, including the freemod variant. `osu-server-spectator` needs to disallow these mods as well.

**Comment by smoogipoo** [comment]
Actually, how about we go the full way and add properties to the `Mod` class itself for these usages? Although we don't yet support custom rulesets, I believe we're going to need this for when we do.

**Comment by bdach** [comment]
> Would probably need to be disabled web-side as well.

For multiplayer yes, but for client submission no. Which is a problem because web doesn't have those paths separated. Not sure this is something I feel comfortable changing on my own without prior discussion of direction.

**Comment by smoogipoo** [comment]
I don't think there's a reason for playlists to disallow any mods other than those that aren't user playable?

File: osu.Game/Rulesets/Mods/IMod.cs [hunk 1]
```diff
@@ -39,6 +39,18 @@ public interface IMod : IEquatable<IMod>
...
+        /// Whether this mod is playable in a multiplayer match.
+        /// Should be <c>false</c> for mods that affect the gameplay progress based on user input (e.g. <see cref="ModAdaptiveSpeed"/>).
+        /// </summary>
+        bool PlayableInMultiplayer { get; }
+
+        /// <summary>
+        /// Whether this mod is valid to be a "free mod" in a multiplayer match.
+        /// Should be <c>false</c> for mods that affect the gameplay progress (e.g. <see cref="ModRateAdjust"/> and <see cref="ModTimeRamp"/>).
+        /// </summary>
+        bool ValidFreeModInMultiplayer { get; }
```
Comment (frenzibyte) [inline osu.Game/Rulesets/Mods/IMod.cs:52]: Was hoping for a better and shorter name than this one (like `PlayableInMultiplayer`), but couldn't come up with anything better than this.

File: osu.Game/Rulesets/Mods/UnknownMod.cs [hunk 2]
```diff
@@ -16,6 +16,7 @@ public class UnknownMod : Mod
         public override double ScoreMultiplier => 0;
 
         public override bool UserPlayable => false;
+        public override bool PlayableInMultiplayer => true;
```
Comment (frenzibyte) [inline osu.Game/Rulesets/Mods/UnknownMod.cs:19]: This override made the naming `PlayableInMultiplayer` fall apart. Thinking about using `SelectableInMultiplayer` or just going the simple way and use `AllowInMultiplayer`.

**Comment by frenzibyte** [comment]
> > [ppy/osu-web@`master`/app/Libraries/Multiplayer/Mod.php](https://github.com/ppy/osu-web/blob/master/app/Libraries/Multiplayer/Mod.php) _looks_ to be exclusively for multiplayer (hence the namespace), but I might be wrong as I'm unfamiliar with the web source code.
> 
> Unfortunately not. [ppy/osu-web@`master`/app/Http/Controllers/Solo/ScoresController.php#L9](https://github.com/ppy/osu-web/blob/master/app/Http/Controllers/Solo/ScoresController.php#L9)

Coming back to this, now that I think about it, disallowing this in web would also prevent using this in playlists mode, and I think https://github.com/ppy/osu-server-spectator/pull/106 should be enough to guard server-side anyways.

File: osu.Game/Rulesets/Mods/UnknownMod.cs (hunk 2 above)
Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/UnknownMod.cs]: Nevermind, this override was just wrong.

File: osu.Game/Utils/ModUtils.cs [hunk 3]
```diff
@@ -112,14 +112,34 @@ public static bool CheckAllowed(IEnumerable<Mod> combination, IEnumerable<Type>
         /// <param name="invalidMods">Invalid mods, if any were found. Can be null if all mods were valid.</param>
         /// <returns>Whether the input mods were all valid. If false, <paramref name="invalidMods"/> will contain all invalid entries.</returns>
         public static bool CheckValidForGameplay(IEnumerable<Mod> mods, [NotNullWhen(false)] out List<Mod>? invalidMods)
+            => checkValid(mods, m => m.Type != ModType.System && m.HasImplementation && !(m is MultiMod), out invalidMods);
+
+        /// <summary>
+        /// Check the provided combination of mods are valid for a multiplayer match session.
+        /// </summary>
+        /// <param name="mods">The mods to check.</param>
+        /// <param name="invalidMods">Invalid mods, if any were found. Can be null if all mods were valid.</param>
```
Comment (bdach) [inline osu.Game/Utils/ModUtils.cs:121]: nit: weird to use "can" in a conditional clause

```suggestion
        /// <param name="invalidMods">Invalid mods, if any were found. Will be null if all mods were valid.</param>
```

goes for the other method too

File: osu.Game/Rulesets/Mods/Mod.cs [hunk 4]
```diff
@@ -94,6 +94,12 @@ public virtual string SettingDescription
         [JsonIgnore]
         public virtual bool UserPlayable => true;
 
+        [JsonIgnore]
+        public virtual bool PlayableInMultiplayer => UserPlayable;
+
+        [JsonIgnore]
+        public virtual bool ValidFreeModInMultiplayer => PlayableInMultiplayer;
```
Comment (bdach) [inline osu.Game/Rulesets/Mods/Mod.cs:101]: These cascading fallbacks make me scared. How am I supposed to tell which value `ValidFreeModInMultiplayer`? Now I have to chase 2 other virtual members by default just to make sure.

Would prefer these to be all explicit and not cascading-order-magic.

File: osu.Game/Rulesets/Mods/IMod.cs [hunk 5]
```diff
@@ -39,6 +39,18 @@ public interface IMod : IEquatable<IMod>
         /// </summary>
         bool UserPlayable { get; }
 
+        /// <summary>
+        /// Whether this mod is playable in a multiplayer match.
+        /// Should be <c>false</c> for mods that affect the gameplay progress based on user input (e.g. <see cref="ModAdaptiveSpeed"/>).
```
Comment (bdach) [inline osu.Game/Rulesets/Mods/IMod.cs:44]:
```suggestion
        /// Should be <c>false</c> for mods that make gameplay duration dependent on user input (e.g. <see cref="ModAdaptiveSpeed"/>).
```

File: osu.Game/Rulesets/Mods/IMod.cs [hunk 6]
```diff
@@ -39,6 +39,18 @@ public interface IMod : IEquatable<IMod>
...
+        /// </summary>
+        bool PlayableInMultiplayer { get; }
+
+        /// <summary>
+        /// Whether this mod is valid to be a "free mod" in a multiplayer match.
+        /// Should be <c>false</c> for mods that affect the gameplay progress (e.g. <see cref="ModRateAdjust"/> and <see cref="ModTimeRamp"/>).
```
Comment (bdach) [inline osu.Game/Rulesets/Mods/IMod.cs:50]:
```suggestion
        /// Should be <c>false</c> for mods that affect the gameplay duration (e.g. <see cref="ModRateAdjust"/> and <see cref="ModTimeRamp"/>).
```

**Review by bdach — DISMISSED** [review]

File: osu.Game/Rulesets/Mods/Mod.cs (hunk 4 above)
Comment (bdach) [reply osu.Game/Rulesets/Mods/Mod.cs:101]: A follow up thought is that with three properties this is likely going to be confusing anyways. Things and places are going to forget to check this.

Suggestion would be to centralise checking mod validity to a single source-of-truth method as follows:

```csharp
public enum ModUsageContext
{
    Solo,
    MultiplayerRequiredMod,
    MultiplayerFreeMod
}

public

File: osu.Game/Utils/ModUtils.cs [hunk 7]
```diff
@@ -112,14 +112,34 @@ public static bool CheckAllowed(IEnumerable<Mod> combination, IEnumerable<Type>
...
+        /// <returns>Whether the input mods were all valid. If false, <paramref name="invalidMods"/> will contain all invalid entries.</returns>
+        public static bool CheckValidFreeModsForMultiplayer(IEnumerable<Mod> mods, [NotNullWhen(false)] out List<Mod>? invalidMods)
+            => checkValid(mods, m => m.ValidFreeModInMultiplayer, out invalidMods);
+
+        private static bool checkValid(IEnumerable<Mod> mods, Predicate<Mod> valid, [NotNullWhen(false)] out List<Mod>? invalidMods)
         {
             mods = mods.ToArray();
-
-            CheckCompatibleSet(mods, out invalidMods);
```
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Utils/ModUtils.cs:118]: I'll probably have to mention it in xmldoc, but with `CheckCompatibleSet`, this can potentially lead to checking compatibility more than once at the same time (see ppy/osu-server-spectator#106).

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Utils/ModUtils.cs:118]: Not super convinced by any of the above. Checking compatibility twice can be avoided server-side by just removing the `CheckCompatibleSet()` call and only leaving the `CheckValidForMultiplayer()` check. `CheckValidFreeModsForMultiplayer()` shouldn't be checking compatibility anyways.

Going to hold out for more opinions on this.

[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Utils/ModUtils.cs:118]: > Not super convinced by any of the above. Checking compatibility twice can be avoided server-side by just removing the CheckCompatibleSet() call and only leaving the CheckValidForMultiplayer() chec

Hard to make that visible by combining them in one exception message, that's exactly the reason I've done that. 

> CheckValidFreeModsForMultiplayer() shouldn't be checking compatibility anyways.

With the change you're proposing, compatibility will be checked in "free mods" context as well.

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Utils/ModUtils.cs:118]: > With the change you're proposing, compatibility will be checked in "free mods" context as well.

Except you don't have to put that `CheckCompatibleSet()` call back into `checkValid()`? That would just be incorrect. The call could however be added to `CheckValidForGameplay()` and `CheckValidForMultiplayer()` one level higher.

File: osu.Game/Rulesets/Mods/Mod.cs (hunk 4 above)
Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/Mod.cs:101]: While this could work in some cases, there are other cases where the caller is shared between solo and multiplayer, e.g.:

https://github.com/ppy/osu/blob/6bf436cd62d0c844d74964837fb2387ebe112725/osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs#L190-L191

In which case we want to ensure the mod is playable at a per-user level, which `Solo` doesn't convey it well. How about calling it `User` instead (and add a xmldoc to it).

Also it slightly feels weird that you could potentially have a mod that has `IsPlayable(ModUsage.MultiplayerRequired) = false` but `IsPlayable(ModUsage.MultiplayerFree) = false`. Would you still be against renaming `MultiplayerRequiredMod` to become just `Multiplayer` in general, and cover `MultiplayerFreeMod` as a result.

Comment (bdach) [reply osu.Game/Rulesets/Mods/Mod.cs:101]: I don't really like `User` as a choice here and the mentioned check could just use `Solo` because the user's own score is what actually matters in the context of that particular call.

Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/Mod.cs:101]: Hmm maybe, let's leave it at `Solo` then.

Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/Mod.cs]: I've pushed the changes so you can look at it and see if it's good as-is.

File: osu.Game/Rulesets/Mods/ModUsage.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+namespace osu.Game.Rulesets.Mods
+{
+    /// <summary>
+    /// The usage of this mod to determine its playability.
+    /// </summary>
+    public enum ModUsage
```
Comment (frenzibyte) [inline osu.Game/Rulesets/Mods/ModUsage.cs:9]: Unsure how to properly document this mod, right now it feels completely off from its purpose (especially for the fact that `Solo` is meant to be for cases where multiplayer is irrelevant, not specifically a solo gameplay session).

Comment (bdach) [reply osu.Game/Rulesets/Mods/ModUsage.cs:9]: Maybe "user" is fine then. Dunno, I wasn't super sure I was going to get buy in from everyone on this idea so pushing it in here already seems rushed.

Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/ModUsage.cs:9]: I think it is a no-brainer to simplify from three cascading properties to one central method, but maybe. I've renamed `Solo` to `User` and adjusted the documentation to hopefully read better.

File: osu.Game/Rulesets/Mods/ModUsage.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+        /// <summary>
+        /// Used as a "required mod" for a multiplayer match.
+        /// </summary>
+        MultiplayerRequired,
+
+        /// <summary>
+        /// Used as a "free mod" for a multiplayer match.
+        /// </summary>
+        MultiplayerFree,
```
Comment (peppy) [inline osu.Game/Rulesets/Mods/ModUsage.cs:24]: This is still very confusing to me. Does `MultiplayerFree` imply that it can also be `MultiplayerRequired`? If so, this needs to be in the naming and documentation.

How about
```csharp
/// Mod can be used in multiplayer but must be applied to all users. This is generally the case for mods which affect the length of gameplay.
MultiplayerRoomWide,
// Mod can be used in multiplayer either at a room or per player level (ie. free mod selection).
MultiplayerPerPlayer,
```
focusing on the actual reasoning behind the limitations.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Mods/ModUsage.cs (hunk 9 above)
Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/ModUsage.cs]: Good suggestion, that looks to work well with the usages of this enum.

File: osu.Game/Rulesets/Mods/Mod.cs [hunk 10]
```diff
@@ -91,10 +91,13 @@ public virtual string SettingDescription
         [JsonIgnore]
         public virtual bool HasImplementation => this is IApplicableMod;
 
+        public virtual bool IsPlayable(ModUsage usage) => true;
+
         [JsonIgnore]
+        [Obsolete("Override IsPlayable instead.")] // Can be removed 20220918
```
Comment (bdach) [inline osu.Game/Rulesets/Mods/Mod.cs:97]: This obsoletion path is really weird because if a consumer has set `UserPlayable => false` then that spec will now be essentially ignored. Usually the new method would delegate to the old in the interim (with a warning suppression) until the consumers have migrated to the new one and it is removed.

**Comment by bdach** [comment]
Are the test failures going to be addressed at some point?

**Comment by frenzibyte** [comment]
Yeah but I've been waiting for opinions on https://github.com/ppy/osu/pull/17270#discussion_r829451080 first, will resolve them for now at least.

File: osu.Game/Rulesets/Mods/Mod.cs (hunk 10 above)
Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/Mod.cs:97]: Right, missed to realize that.

File: osu.Game/Utils/ModUtils.cs [hunk 11]
```diff
@@ -106,20 +106,40 @@ public static bool CheckAllowed(IEnumerable<Mod> combination, IEnumerable<Type>
...
         /// <summary>
-        /// Check the provided combination of mods are valid for a local gameplay session.
+        /// Checks that all <see cref="Mod"/>s in a combination are valid for a local gameplay session.
         /// </summary>
         /// <param name="mods">The mods to check.</param>
-        /// <param name="invalidMods">Invalid mods, if any were found. Can be null if all mods were valid.</param>
+        /// <param name="invalidMods">Invalid mods, if any were found. Will be null if all mods were valid.</param>
         /// <returns>Whether the input mods were all valid. If false, <paramref name="invalidMods"/> will contain all invalid entries.</returns>
         public static bool CheckValidForGameplay(IEnumerable<Mod> mods, [NotNullWhen(false)] out List<Mod>? invalidMods)
+            => checkValid(mods, m => m.Type != ModType.System && m.HasImplementation && !(m is MultiMod), out invalidMods);
```
Comment (smoogipoo) [inline osu.Game/Utils/ModUtils.cs:115]: Is there a reason why this isn't going through `IsPlayable(User)`?

File: osu.Game/Rulesets/Mods/ModUsage.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace osu.Game.Rulesets.Mods
+{
+    /// <summary>
+    /// The usage of this mod to determine whether it's playable in such context.
+    /// </summary>
+    public enum ModUsage
+    {
+        /// <summary>
+        /// Used for a per-user gameplay session.
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Mods/ModUsage.cs:12]:
```suggestion
        /// Used for a solo gameplay session.
```

File: osu.Game/Rulesets/Mods/ModUsage.cs [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+        /// Determines whether the mod is playable by an end user.
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Mods/ModUsage.cs:13]: This line's redundant, suggest removing

File: osu.Game/Rulesets/Mods/ModUsage.cs [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+        /// </summary>
+        User,
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Mods/ModUsage.cs:15]: Maybe `Local`?

File: osu.Game/Rulesets/Mods/ModUsage.cs [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+        /// <summary>
+        /// Used in multiplayer but must be applied to all users.
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Mods/ModUsage.cs:18]: I suggest renaming these xmldocs to read more like: `The mod can be used in multiplayer but ...`

File: osu.Game/Rulesets/Mods/ModUsage.cs [hunk 16]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+namespace osu.Game.Rulesets.Mods
+{
+    /// <summary>
+    /// The usage of this mod to determine whether it's playable in such context.
+    /// </summary>
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Mods/ModUsage.cs:8]: How about `The context in which a <see cref="Mod"/> is playable.`?

File: osu.Game.Tests/Mods/ModUtilsTest.cs [hunk 17]
```diff
@@ -134,18 +134,6 @@ public void TestNotAllowedThroughBaseType()
...
-            {
-                new Mod[] { new OsuModDoubleTime(), new OsuModHalfTime() },
-                new[] { typeof(OsuModDoubleTime), typeof(OsuModHalfTime) }
-            },
-            // incompatible pair with derived class.
-            new object[]
-            {
-                new Mod[] { new OsuModNightcore(), new OsuModHalfTime() },
-                new[] { typeof(OsuModNightcore), typeof(OsuModHalfTime) }
-            },
```
Comment (smoogipoo) [inline osu.Game.Tests/Mods/ModUtilsTest.cs:148]: Why were these removed?

**Review by smoogipoo — CHANGES_REQUESTED** [review]
Generally, I think I'm fine with this. Would also require a review from @peppy.

Not sure on the naming of the enum, my suggestion is:
```
SoloLocal,
MultiplayerGlobal,
MultiplayerLocal
```

File: osu.Game/Utils/ModUtils.cs (hunk 11 above)
Comment (frenzibyte) [reply osu.Game/Utils/ModUtils.cs:115]: It wasn't checked for previously, but can add it there.

**Comment by frenzibyte** [comment]
> Generally, I think I'm fine with this. Would also require a review from @peppy.
> 
> Not sure on the naming of the enum, my suggestion is:
> 
> ```
> SoloLocal,
> MultiplayerGlobal,
> MultiplayerLocal
> ```

Unsure how to feel about the term "multiplayer local" for free mods, but that might just be me. Would like to know more opinions on this.

**Comment by peppy** [comment]
"Local" seems redundant and only adds confusion in "SoloLocal".

I also prefer the naming I proposed over "Global" and "Local", as these still require attached xmldoc to explain what they mean.

**Comment by peppy** [comment]
I'm okay to go with the proposed naming if it's blocking progress on this PR.

As an aside, on a re-read is almost looks like `ModUsage` should be a bitwise enum and mods would specify which modes they allow. But maybe it's fine as is..

**Comment by frenzibyte** [comment]
Making it a bitwise event wouldn't exactly work since one enum member could indicate more at once, unless you mean something like this:
```csharp
        /// <summary>
        /// This mod can be used for a per-user gameplay session.
        /// </summary>
        SoloLocal = 1,

        /// <summary>
        /// This mod can be used in multiplayer but must be applied to all users.
        /// This is generally the case for mods which affect the length of gameplay.
        /// </summary>
        MultiplayerGlobal = SoloLocal | 2,

        /// <summary>
        /// This mod can be used in multiplayer either at a room or per-player level (i.e. "free mod").
        /// </summary>
        MultiplayerLocal = MultiplayerGlobal | 4,
```
which conveys the rule that a mod playable at `MultiplayerGlobal` means playable at `SoloLocal`, but not the other way around.

**Comment by bdach** [comment]
As for making `ModUsage` a flags enum, I'm a bit wary of combining flags [as proposed above](https://github.com/ppy/osu/pull/17270#issuecomment-1117385209) as it introduces some implicit assumptions which I don't know if they will hold up in the future (as in, I'm not sure if multiplayer-only mods are going to ever exist), so I'd rather stay loose with this API for now and only tighten it down if it turns out to be an issue.

File: osu.Game/Utils/ModUtils.cs [hunk 18]
```diff
@@ -121,7 +130,14 @@ public static bool CheckValidForGameplay(IEnumerable<Mod> mods, [NotNullWhen(fal
         /// <param name="invalidMods">Invalid mods, if any were found. Will be null if all mods were valid.</param>
         /// <returns>Whether the input mods were all valid. If false, <paramref name="invalidMods"/> will contain all invalid entries.</returns>
         public static bool CheckValidRequiredModsForMultiplayer(IEnumerable<Mod> mods, [NotNullWhen(false)] out List<Mod>? invalidMods)
-            => checkValid(mods, m => m.IsPlayable(ModUsage.MultiplayerGlobal), out invalidMods, true);
+        {
+            mods = mods.ToArray();
+
+            if (!CheckCompatibleSet(mods, out invalidMods))
```
Comment (bdach) [inline osu.Game/Utils/ModUtils.cs:136]: Not sure why this isn't handling `MultiMod`s while `CheckValidForGameplay()` is. The two should match in behaviour.

I can see two ways to go about it: either mirror the code in `CheckValidForGameplay()`, or split a separate `checkValid(mods, m => !(m is MultiMod), out invalidMods);` call preceding this in both places.

File: osu.Game/Utils/ModUtils.cs [hunk 19]
```diff
@@ -112,7 +112,16 @@ public static bool CheckAllowed(IEnumerable<Mod> combination, IEnumerable<Type>
         /// <param name="invalidMods">Invalid mods, if any were found. Will be null if all mods were valid.</param>
         /// <returns>Whether the input mods were all valid. If false, <paramref name="invalidMods"/> will contain all invalid entries.</returns>
         public static bool CheckValidForGameplay(IEnumerable<Mod> mods, [NotNullWhen(false)] out List<Mod>? invalidMods)
-            => checkValid(mods, m => m.Type != ModType.System && m.HasImplementation && !(m is MultiMod), out invalidMods, true);
+        {
+            mods = mods.ToArray();
+
+            // exclude multi mods from compatibility checks.
+            // the loop below automatically marks all multi mods as not valid for gameplay anyway.
```
Comment (bdach) [inline osu.Game/Utils/ModUtils.cs:119]:
```suggestion
            // the `checkValid()` call below marks all multi mods as not valid for gameplay anyway.
```

But this may potentially change, depending on what happens with the other review conversation (potentially it will be "the `checkValid()` call above").

File: osu.Game/Utils/ModUtils.cs (hunk 18 above)
Comment (frenzibyte) [reply osu.Game/Utils/ModUtils.cs:136]: Hmm, thought it's pointless to check that in `CheckValidForMultiplayerRequiredMods` since it shouldn't even be a valid in the game-wide mods bindable, but I guess that's irrelevant in this case as this could get sent directly to the server. Replacing with `!(m is MultiMod)` makes sense.

File: osu.Game/Utils/ModUtils.cs (hunk 19 above)
Comment (frenzibyte) [reply osu.Game/Utils/ModUtils.cs]: As per the suggestion in the other review comment, I've moved this comment to the `checkValid` method and briefly explained why it's separated from the post-compatibility one.

File: osu.Game/Rulesets/Mods/IMod.cs [hunk 20]
```diff
@@ -33,10 +33,24 @@ public interface IMod : IEquatable<IMod>
         /// </summary>
         IconUsage? Icon { get; }
 
+        /// <summary>
+        /// Whether this mod is playable for the given usage.
```
Comment (peppy) [inline osu.Game/Rulesets/Mods/IMod.cs:37]: This is missing context which results in the purpose of the method being undefined. Is the end purpose of this method in line with the obsoleted one? ie. can we rename it to `RequiresUserInteraction`? Otherwise I'm not sure how a newcomer to the project could make sense of this.

If `IsPlayable(SoloLocal)` is false, a mod is still selectable during solo local, but the mod is non-interactive.

If `IsPlayable(MultiplayerLocal)` (i still cannot get behind this "local" naming, but whatever) is false, the mod should never be shown anywhere.

Is this correct thinking? If so, I have a hard time thinking of how to phrase this xmldoc, but what is currently written definitely doesn't do a good job. At least the obsoleted version mentioned "playable by an end user", which explains what "playable" actually means here.

Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/IMod.cs:37]: I think the problem here is that the way for checking whether a mod is selectable at multiplayer was fused with `UserPlayable` with the naming, which is completely different from what the method is meant to achieve.

One final suggestion I have is to rename it to `IsValidFor(ModUsage usage)`, or stick back to the old way of having two non-cascading boolean properties (`ValidForMultiplayer`, `ValidForMultiplayerAsFreeMod`) and call it a day.

@bdach need your opinion on this, or if you have a better idea to improve this.

Comment (peppy) [reply osu.Game/Rulesets/Mods/IMod.cs:37]: I'd prefer two separate `bool`s, personally. If we are sure that these are the two well-defined usages we need. The naming you suggest seems fine.

Comment (bdach) [reply osu.Game/Rulesets/Mods/IMod.cs:37]: I don't know, I kinda give up. I can live with three boolean flags, so long as they are *separated* from each other and don't fall back to each other via virtual properties like they did before. I hope I'm not the only one who thinks that having three overridable virtual properties check each other in sequence is a guaranteed way to produce some unholy abomination sometime down the line.

Comment (peppy) [reply osu.Game/Rulesets/Mods/IMod.cs:37]: Is there a usage I'm missing with sequential checks, or are you referring to chaining the values?

The main issue I have is the washy definition of what "playable" is, and how autoplay almost feels like an edge case here.

Comment (bdach) [reply osu.Game/Rulesets/Mods/IMod.cs:37]: I'm referring to the chaining alone. I don't want to see stuff like this, essentially:

image

which is what this PR initially had.

Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/IMod.cs:37]: Yeah, can agree with not chaining, don't think there's anything that relies on that anyways.

File: osu.Game/Rulesets/Mods/ModAutoplay.cs [hunk 21]
```diff
@@ -25,6 +25,8 @@ public abstract class ModAutoplay : Mod, IApplicableFailOverride, ICreateReplayD
         public bool RestartOnFail => false;
 
         public override bool UserPlayable => false;
+        public override bool ValidForMultiplayer => false;
+        public override bool ValidForMultiplayerAsFreeMod => false;
```
Comment (frenzibyte) [inline osu.Game/Rulesets/Mods/ModAutoplay.cs:29]: This may be one down from using separate boolean properties (having to specify each and every one to make sense), maybe we can enforce validity of those via `Debug.Assert` in `Mod` constructor? not sure if that's pleasant at all though.

**Comment by bdach** [comment]
Aside from the above I can live with this diff as is currently.

File: osu.Game/Utils/ModUtils.cs [hunk 22]
```diff
@@ -106,22 +106,64 @@ public static bool CheckAllowed(IEnumerable<Mod> combination, IEnumerable<Type>
...
+                return false;
+
+            if (!CheckCompatibleSet(mods, out invalidMods))
+                return false;
+
+            return checkValid(mods, m => m.Type != ModType.System && m.HasImplementation && m.ValidForMultiplayer, out invalidMods);
+        }
+
+        /// <summary>
+        /// Checks that all <see cref="Mod"/>s in a combination are valid as "free mods" in a multiplayer match session.
```
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Utils/ModUtils.cs:151]: Hmm, I would usually think "free mods" are the mods allowed in the room, while "user mods" are the mods the user currently selected from the allowed list. Just mentioning that compatibility is not checked with a bit of context about that sounds fair.

**Review by peppy — APPROVED** [review]

**Review by smoogipoo — APPROVED** [review]
